src/emulator/kernel/src/thread.cpp

```diff
@@ -16,16 +16,45 @@
 // 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 
 #include <kernel/thread_functions.h>
-
 #include <kernel/thread_state.h>
+#include <psp2/types.h>
+#include <psp2/kernel/error.h>
+
+#include <kernel/state.h>
 
 #include <cpu/functions.h>
 #include <util/resource.h>
+#include <util/lock_and_find.h>
+#include <util/find.h>
+
+#include <SDL_thread.h>
 
 #include <cassert>
 #include <cstring>
 
-ThreadStatePtr init_thread(Ptr<const void> entry_point, size_t stack_size, bool log_code, MemState &mem, CallImport call_import) {
+struct ThreadParams {
+    KernelState *kernel = nullptr;
+    SceUID thid = SCE_KERNEL_ERROR_ILLEGAL_THREAD_ID;
+    SceSize arglen = 0;
+    Ptr<void> argp;
+    std::shared_ptr<SDL_semaphore> host_may_destroy_params = std::shared_ptr<SDL_semaphore>(SDL_CreateSemaphore(0), SDL_DestroySemaphore);
+};
+
+static int SDLCALL thread_function(void *data) {
+    assert(data != nullptr);
+    const ThreadParams params = *static_cast<const ThreadParams *>(data);
+    SDL_SemPost(params.host_may_destroy_params.get());
+    const ThreadStatePtr thread = lock_and_find(params.thid, params.kernel->threads, params.kernel->mutex);
+    write_reg(*thread->cpu, 0, params.arglen);
+    write_reg(*thread->cpu, 0, params.argp.address());
+    const bool succeeded = run_thread(*thread);
+    assert(succeeded);
+    const uint32_t r0 = read_reg(*thread->cpu, 0);
+    return r0;
+}
+
+typedef std::function<void(uint32_t)> CallImportForThread;
+ThreadStatePtr init_thread(Ptr<const void> entry_point, size_t stack_size, bool log_code, MemState &mem, CallImportForThread call_import) {
```

Comment:
Is `init_thread` called anywhere other than `create_thread`?

If not, I would probably combine them into one function. That may help reduce the need for an extra `CallImportForThread ` type too.

Reply:
It's not called anymore so I'll combine them